# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [1]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric

In [17]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [3]:
x.shape

(506, 13)

In [4]:
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

<br><br><br>

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/BTVYXQ1.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [5]:
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 
    
    row, column = input_data.shape
    selecting_rows = np.random.choice(list(range(row)), size = 303, replace = False)
    replacing_rows = np.random.choice(list(range(len(selecting_rows))), size = 203, replace = True)
    col_size = np.random.choice(list(range(column)), size = 1, replace = True)
    selecting_columns = np.random.choice(column, size = col_size+1, replace = False)

    # source: https://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.ix_.html
    sample_data = input_data[np.ix_(selecting_rows,selecting_columns)]
    target_sample_data = target_data[selecting_rows]


    #replicating data
    replicated_sample_data = sample_data[replacing_rows]
    target_of_replicated_sample_data = target_sample_data[replacing_rows]

    #concatenating data
    final_sample_data = np.vstack((sample_data,replicated_sample_data))
    final_target_data = np.vstack((target_sample_data.reshape(-1,1),target_of_replicated_sample_data.reshape(-1,1)))

    return final_sample_data,final_target_data,selecting_rows,selecting_columns

    # return sampled_input_data , sampled_target_data,selecting_rows,selecting_columns
    #note please return as lists

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [6]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [7]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]

for i in range(0,30):
    a,b,c,d = generating_samples(x,y)
    list_input_data.append(a)
    list_output_data.append(b)
    list_selected_row.append(c)
    list_selected_columns.append(d)

<font color='cyan'> <b>Grader function - 2 </b></font>

In [8]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [9]:
from sklearn.tree import DecisionTreeRegressor
list_of_all_models = []
array_of_y = []
for i in range (0,30):
    model = DecisionTreeRegressor(max_depth = None, random_state = 0)
    model.fit(list_input_data[i],list_output_data[i])
    y = model.predict(x[:,list_selected_columns[i]]).reshape(1,-1)
    list_of_all_models.append(model)
    array_of_y.append(y)

In [10]:
len(array_of_y)

30

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

In [11]:
predicted_y = np.median(array_of_y, axis = 0)
predicted_y

array([[24.  , 21.6 , 34.7 , 33.4 , 36.2 , 28.7 , 22.9 , 21.35, 16.5 ,
        18.9 , 15.  , 18.9 , 21.7 , 20.4 , 18.2 , 19.9 , 23.1 , 17.5 ,
        20.2 , 18.2 , 13.6 , 19.6 , 15.2 , 14.5 , 15.6 , 13.9 , 16.6 ,
        14.8 , 18.4 , 21.  , 12.7 , 14.5 , 13.2 , 13.1 , 13.5 , 18.9 ,
        20.  , 21.  , 24.7 , 30.8 , 34.9 , 26.6 , 25.3 , 24.7 , 21.2 ,
        19.3 , 20.  , 16.6 , 14.4 , 19.4 , 19.7 , 20.5 , 25.  , 23.4 ,
        18.9 , 35.4 , 24.7 , 31.6 , 23.3 , 19.6 , 18.7 , 16.  , 22.2 ,
        25.  , 33.  , 23.5 , 19.4 , 22.  , 17.6 , 20.9 , 24.2 , 21.7 ,
        22.8 , 23.4 , 24.1 , 21.4 , 20.  , 20.8 , 21.2 , 20.3 , 28.  ,
        23.9 , 24.8 , 22.9 , 23.9 , 26.6 , 22.5 , 22.2 , 23.6 , 28.7 ,
        22.6 , 22.  , 22.9 , 25.  , 20.6 , 28.4 , 21.4 , 38.7 , 43.8 ,
        33.2 , 27.5 , 26.5 , 18.6 , 19.3 , 20.1 , 19.5 , 19.5 , 20.4 ,
        19.8 , 19.4 , 21.7 , 22.8 , 18.8 , 18.7 , 18.5 , 18.3 , 21.2 ,
        19.2 , 20.4 , 19.3 , 22.  , 20.3 , 20.5 , 17.3 , 18.8 , 20.85,
      

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [13]:
from sklearn.metrics import mean_squared_error
mean_squared_error(y, predicted_y)

11.584140316205533

<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [14]:
from tqdm import tqdm
y_pred = []
for datapoint in tqdm(range (0,len(x))):
    list_of_y_values = []
    for i,sample in enumerate(list_selected_row):
        if datapoint not in sample:
            xq = x[datapoint,list_selected_columns[i]]
            list_of_y_values.append(list_of_all_models[i].predict(xq.reshape(1,-1)))
    pred = np.median(list_of_y_values)
    y_pred.append(pred)

100%|██████████| 506/506 [00:00<00:00, 1041.02it/s]


In [15]:
len(y)

1

In [18]:
oob_score = mean_squared_error(y, y_pred)

In [19]:
oob_score

18.063606129734616

# <font color='blue'><b>Task 2</b></font>

In [39]:
from tqdm import tqdm
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error


oob_score = []
mse = []

for train in tqdm(range(0,35)):


    x=boston.data #independent variables
    y=boston.target #target variable
    
    list_input_data =[]
    list_output_data =[]
    list_selected_row= []
    list_selected_columns=[]
    
    for i in range(0,30):
        a,b,c,d = generating_samples(x,y)
        list_input_data.append(a)
        list_output_data.append(b)
        list_selected_row.append(c)
        list_selected_columns.append(d)
    
    predicted_y = 0
    list_of_all_models = []
    array_of_y = []
    for i in range (0,30):
        model = DecisionTreeRegressor(max_depth = None, random_state = 0)
        model.fit(list_input_data[i],list_output_data[i])
        y = model.predict(x[:,list_selected_columns[i]]).reshape(1,-1)
        list_of_all_models.append(model)
        array_of_y.append(y)
    
    predicted_y = np.median(array_of_y, axis = 0)

    mse.append(mean_squared_error(y, predicted_y))


    y_pred = []
    pred = 0
    for datapoint in range (0,len(x)):
        list_of_y_values = []
        for i,sample in enumerate(list_selected_row):
            if datapoint not in sample:
                xq = x[datapoint,list_selected_columns[i]]
                list_of_y_values.append(list_of_all_models[i].predict(xq.reshape(1,-1)))
        pred = np.median(list_of_y_values)
        y_pred.append(pred)
    oob_score.append(mean_squared_error(boston.target, y_pred))


100%|██████████| 35/35 [00:18<00:00,  1.90it/s]


In [40]:
mse

[34.62684357439051,
 6.609801400316861,
 21.9049917568235,
 5.7222189831293475,
 12.041892841458058,
 12.4138353219767,
 16.68857213438735,
 10.938722995132522,
 49.84616583211693,
 8.688328571400021,
 9.317925305748334,
 33.27544827647387,
 8.395282555514383,
 8.139421814766017,
 12.101765437325854,
 52.71061014640316,
 7.733821787062146,
 11.721902378984382,
 10.403496437808032,
 31.030147294352158,
 13.76820144883349,
 26.59986430835076,
 24.99911932768529,
 10.190951843900711,
 12.18833511747914,
 17.48402065088356,
 51.69335119496513,
 12.574066780729288,
 17.076170020574363,
 8.965331027667983,
 35.49224076389449,
 7.672149286706483,
 6.685783526268853,
 61.85897413950919,
 16.49765096362836]

In [42]:
oob_score

[13.888954511716078,
 18.409170479320036,
 18.050154404673968,
 15.886193466807015,
 16.099757488843665,
 15.421374971082589,
 13.925728727492313,
 19.807664047436198,
 14.978781240498027,
 16.89800891381254,
 16.192496261287292,
 19.087099780200766,
 17.936759804826135,
 14.493082284138918,
 14.430957911215325,
 12.556028248350477,
 19.092240581656256,
 14.267884740187068,
 16.164905654937336,
 18.588236944996225,
 14.529392764359438,
 14.639604257891234,
 17.63346668449696,
 16.167893273920445,
 15.4259801714958,
 14.210057036559228,
 17.566258148426108,
 14.959250406430813,
 12.466895334357014,
 18.88556426904919,
 15.984796743864493,
 17.591712613718656,
 16.36211938158161,
 17.2865614372077,
 18.782638233892246]

In [47]:
# CI for MSE
MSE_left_limit  = np.round(np.mean(mse) - 2*(np.std(mse)/np.sqrt(35)), 3)
MSE_right_limit = np.round(np.mean(mse) + 2*(np.std(mse)/np.sqrt(35)), 3)
print("Confidence Interval for MSE is [{0},{1}]".format(MSE_left_limit,MSE_right_limit))

OOB_left_limit  = np.round(np.mean(oob_score) - 2*(np.std(oob_score)/np.sqrt(35)), 3)
OOB_right_limit = np.round(np.mean(oob_score) + 2*(np.std(oob_score)/np.sqrt(35)), 3)
print("Confidence Interval for Out-of-Bag scores is [{0},{1}]".format(MSE_left_limit,MSE_right_limit))

Confidence Interval for MSE is [14.627,24.691]
Confidence Interval for Out-of-Bag scores is [14.627,24.691]


# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [20]:
xq = [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60]
y_xq = []
for i in range(0,len(list_of_all_models)):
    Xq = np.array(xq).reshape(-1,1)
    y_xq.append(list_of_all_models[i].predict(Xq[list_selected_columns[i]].reshape(1,-1)))
predict_Y_of_datapoint =np.median(y_xq)
print(predict_Y_of_datapoint)

19.75


<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>

For task 1, we can see that the MSE is pretty small. It is caused by the tree that was built on the training sit to the maximum depth, thus it will overfit to the training data. 

For task 2, we can see that after we run the random forest for 35 times, we can find the confidence interval. The Confidence Interval for MSE is [14.627,24.691] and the Confidence Interval for Out-of-Bag scores is [14.627,24.691]

For task 3, given the data point xq, our model is able to predict the price to be 19.75.
